### PR TITLE
[clang][ObjectiveC] Fix Parsing the `::` Optional Scope Specifier

### PR DESCRIPTION
**Pull request by qiongsiwu**
The parser hangs when processing types/variables prefixed by `::` as an optional scope specifier. For example, 
```
- (instancetype)init:(::A *) foo;
```

The parser should not hang, and it should emit an error. This PR implements the error check. 

rdar://140885078

### CODE REVIEW

**Comment by qiongsiwu** [comment]
This fix tripped some tests in `libc++`. I am investigating.

**Review by shafik — COMMENTED** [review]
This should have a release note.

**Comment by vsapsai** [comment]
How does it work in Objective-C++? I don't know even if we have a test but hope we do.

**Comment by cor3ntin** [comment]
> How does it work in Objective-C++? I don't know even if we have a test but hope we do.

Right, this needs tests for Objective-C++

**Comment by vsapsai** [comment]
I don't know if we have a test for it but I've realized there are cases where you can have a legitimate double colon in Objective-C. For example,
```objective-c
@interface NSObject
@end

@implementation NSObject
- (void)performSelector:(SEL)selector {}

- (void)double:(int)firstArg :(int)secondArg colon:(int)thirdArg {}

- (void)test {
  [self performSelector:@selector(double::colon:)];
}
@end
```

It's not a method parameter type, so it is possible your code isn't executed. But it is worth checking if we test this case.

File: clang/lib/Parse/Parser.cpp [hunk 1]
```diff
@@ -2222,8 +2222,15 @@ bool Parser::TryAnnotateTypeOrScopeTokenAfterScopeSpec(
     }
   }
 
-  if (SS.isEmpty())
+  if (SS.isEmpty()) {
+    if (getLangOpts().ObjC && !getLangOpts().CPlusPlus &&
```
[the remark this answers is not in the document]
Comment (qiongsiwu) [reply clang/lib/Parse/Parser.cpp:2226]: Good question. The reason I added the guard for Objective C is a test case failure if we only rule out `CPlusPlus`. This test case https://github.com/qiongsiwu/llvm-project/blob/139e69b7bcb05e6ff9db0f373d9180deb341a571/clang/test/ParserOpenACC/parse-clauses.c#L436 uses `::` for OpenACC.

**Comment by qiongsiwu** [comment]
Thanks for the feedback everyone! The PR is updated to address the review comments. 

- A test is added to cover Objective-C++. https://github.com/llvm/llvm-project/pull/119908/files#diff-7f1cce2d0397139fc73d3615c8249ac9530082aa707868c0e921f53daf594384R1 I am not sure if this is sufficient so please provide feedback. 
- The Objective-C test is revised to cover the optional scope specifier on variables (in addition to parameters) https://github.com/llvm/llvm-project/pull/119908/files#diff-0c0e8844905a75c581903811a694e614f49b2945e2254a76bb5a215443028554R11. 
- The release note is updated https://github.com/llvm/llvm-project/pull/119908/files#diff-ec770381d76c859f5f572db789175fe44410a72608f58ad5dbb14335ba56eb97R678.

**Comment by qiongsiwu** [comment]
The Objective-C test is modified to contain valid code with `::` to include @vsapsai 's suggestion . https://github.com/llvm/llvm-project/pull/119908/files#diff-0c0e8844905a75c581903811a694e614f49b2945e2254a76bb5a215443028554R11

**Comment by qiongsiwu** [comment]
Gentle ping for review. Thanks so much!

**Review by cyndyishida — APPROVED** [review]

**Comment by MaskRay** [comment]
(https://discourse.llvm.org/t/hidden-emails-on-github-should-we-do-something-about-it/74223 the convention is to disable github's hide-email feature)

**Pull request merged**